File: Algorithm_LevelOne/Letter_Case_Permutation/Letter_Case_Permutation.py
# 784. Letter Case Permutation Python

class Solution(object):
    def letterCasePermutation(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        output = [[]]

        for char in s:
            n = len(output)
            if char.isalpha():
                for i in range(n):
                    output.append(output[i][:])
                    output[i].append(char.lower())
                    output[n+i].append(char.upper())
            else:
                for i in range(n):
                    output[i].append(char)
        return map("".join, output)

File: Algorithm_LevelOne/Letter_Case_Permutation/Letter_Case_Permutation.cc
// 784. Letter Case Permutation C++

#include <string>
#include <vector>
class Solution {
public:
    std::vector<std::string> letterCasePermutation(std::string s) {
        std::vector<std::string> ans;
        backtrack(s, ans, 0);
        return ans;
    }
    void backtrack(std::string s, std::vector<std::string>& ans, int i) {
        if (i == s.size()) {
            ans.emplace_back(s);
            return;
        }

        if (isalpha(s[i])) {
            s[i] = toupper(s[i]);
            backtrack(s,ans,i+1);
            s[i] = tolower(s[i]);
            backtrack(s,ans,i+1);
        } else {
            backtrack(s, ans, i+1);
        }
    }
};
